sqlmesh/core/metric/definition.py

```diff
@@ -0,0 +1,193 @@
+from sqlmesh.utils import UniqueKeyDict
+from sqlmesh.utils.errors import ConfigError
+from sqlmesh.utils.pydantic import PydanticModel
+
+
+def load_metric_ddl(
+    expression: exp.Expression, dialect: t.Optional[str], path: Path = Path(), **kwargs: t.Any
+) -> MetricMeta:
+    """Returns a MetricMeta from raw Metric DDL."""
+    if not isinstance(expression, d.Metric):
+        _raise_metric_config_error(
+            f"Only METRIC(...) statements are allowed. Found {expression.sql(pretty=True)}", path
+        )
+
+    metric = MetricMeta(
+        **{
+            "dialect": dialect,
+            "description": "\n".join(comment.strip() for comment in expression.comments)
+            if expression.comments
+            else None,
+            **{prop.name.lower(): prop.args.get("value") for prop in expression.expressions},
+            **kwargs,
+        }
+    )
+
+    metric._path = path
+
+    return metric
+
+
+def expand_metrics(metas: UniqueKeyDict[str, MetricMeta]) -> UniqueKeyDict[str, Metric]:
+    """Resolves all metas into standalone metrics."""
+    metrics: UniqueKeyDict[str, Metric] = UniqueKeyDict("metrics")
+
+    for name, meta in metas.items():
+        if name not in metrics:
+            metrics[name] = meta.to_metric(metas, metrics)
+
+    return metrics
+
+
+def find_tables(expression: exp.Expression, dialect: str) -> t.Set[str]:
+    """Finds all the table references in a metric definition."""
+    return {
+        d.normalize_model_name(exp.table_(*reversed(column.parts[:-1])), dialect=dialect)  # type: ignore
+        for column in expression.find_all(exp.Column)
+        if column.table
+    }
+
+
+@t.overload
+def remove_namespace(expression: str) -> str:
+    ...
+
+
+@t.overload
+def remove_namespace(expression: exp.Column, dialect: str) -> str:
+    ...
+
+
+def remove_namespace(expression: str | exp.Column, dialect: t.Optional[str] = None) -> str:
+    """Given a column or a string, rewrite table namespaces like catalog.db to catalog__db"""
+
+    if not isinstance(expression, str):
+        assert dialect is not None
+        expression = first(find_tables(expression, dialect=dialect))
+    return expression.replace(".", "__")
+
+
+class MetricMeta(PydanticModel, frozen=True):
+    """Raw metric definition without relationships or expansion of derived metrics."""
+
+    name: str
+    dialect: str
+    expression: exp.Expression
+    description: t.Optional[str] = None
+    owner: t.Optional[str] = None
+
+    _path: Path = Path()
+
+    @validator("name", pre=True)
+    def _name_validator(cls, v: t.Any, values: t.Dict[str, t.Any]) -> str:
+        return cls._string_validator(v).lower()
+
+    @validator("dialect", "owner", "description", pre=True)
+    def _string_validator(cls, v: t.Any) -> t.Optional[str]:
+        if isinstance(v, exp.Expression):
+            return v.name
+        return str(v) if v is not None else None
+
+    @validator("expression", pre=True)
+    def _validate_expression(
```

Comment:
https://github.com/TobikoData/sqlmesh/blob/main/sqlmesh/core/model/common.py#L46

Reply:
i purposefully didn't use this one because it seemed more complicated but i guess i could